add JSON processing capabilities to sensor_serial

dgomes opened the pull request:
## Description:

Reading from the serial line strings/numbers is OK, but most likely you will be reading something more complex. This pull request address devices that publish in JSON format through the serial interface adding the value_template optional configuration (which exists in other sensors) and copying into attributes the JSON values.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3942

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

dgomes committed:
add JSON processing capabilities

dgomes committed:
format fixes

dgomes committed:
format fixes

fabaff commented on homeassistant/components/sensor/serial.py:
```diff
@@ -71,7 +80,18 @@ def serial_read(self, device, rate, **kwargs):
+            try:
+                self._attributes = json.loads(line)
+            except json.JSONDecodeError:
+                self._attributes = []
```
This doesn't make much sense. There is no reason to have the state duplicated into an attribute if only one value, the state, is available. Beside that will this lead to a `TypeError` for the attributes. 

Something like this could work if only one value is received:

```python
            try:
                data = json.loads(line)
            except json.JSONDecodeError:
                _LOGGER.error("Invalid JSON data received: %s", data)

            if isinstance(data, dict):
                self._attributes = data
```

dgomes replied:
the except makes no sense, since when you only have a string it will necessary log an error (which is not a real error).

so now I just pass the exception altogether ignoring when it's not a JSON

fabaff replied:
Well, a string is not valid JSON.

Being for tolerant for data that is coming from a device is always good.

fabaff commented:
This also requires an update of the existing documentation.

dgomes committed:
Fix according to @fabaff comment

dgomes commented:
Well, someone updated the documentation already :) 

or am I missing something ?

fabaff commented:
```bash
./homeassistant/components/sensor/serial.py:117:1: D202 No blank lines allowed after function docstring
./homeassistant/components/sensor/serial.py:117:1: D208 Docstring is over-indented
```

dgomes committed:
reverting last commit to a more sane approach

dgomes committed:
docstring...

dgomes committed:
still docstring...

dgomes committed:
passed script/lint

dgomes committed:
Merge remote-tracking branch 'home-assistant/dev' into serial_JSON

dgomes committed:
downgrade exception

JSONDecodeError was only introduced in Python3.5

Since we are still supporting 3.4 ValueError is the parent class of
JSONDecodeError

MartinHjelmare commented on homeassistant/components/sensor/serial.py:
```diff
@@ -91,6 +113,11 @@ def should_poll(self):
         return False
 
     @property
+    def state_attributes(self):
```
Don't overwrite `state_attributes`, use `device_state_attributes`.

dgomes replied:
Can't you elaborate on that ? Or point me towards some example/documentation on the differences?

I basically followed the REST sensor example, but I'm available to improve the code.

MartinHjelmare replied:
The former should only be implemented by base components, the latter by platforms.

https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py#L104-L117

dgomes replied:
I get it, but it's currently just semantics. The pull request is closed, can any of view just make the change :) ?

MartinHjelmare replied:
Just make a new PR with the change and we can quickly merge it.
